dhall/src/Dhall/Eval.hs

```diff
@@ -0,0 +1,900 @@
+                                    as
+                            in VListLit a' as'
+                          t -> VListIndexed a t
+    ListReverse      -> VPrim $ \ ~a -> VPrim $ \case
+                          VListLit t as | null as -> VListLit t (Data.Sequence.reverse as)
+                          VListLit t as -> VListLit Nothing (Data.Sequence.reverse as)
+                          t             -> VListReverse a t
+
+    Optional         -> VPrim VOptional
+    OptionalLit a mt -> maybe (VNone (evalE a)) (\t -> VSome (evalE t)) mt
+    Some t           -> VSome (evalE t)
+    None             -> VPrim $ \ ~a -> VNone a
+
+    OptionalFold     -> VPrim $ \ ~a -> VPrim $ \case
+                          VNone _ ->
+                            VHLam (Typed "optional" (VConst Type)) $ \optional ->
+                            VHLam (Typed "some" (vFun a optional)) $ \some ->
+                            VHLam (Typed "none" optional) $ \none ->
+                            none
+                          VSome t ->
+                            VHLam (Typed "optional" (VConst Type)) $ \optional ->
+                            VHLam (Typed "some" (vFun a optional)) $ \some ->
+                            VHLam (Typed "none" optional) $ \none ->
+                            some `vApp` t
+                          opt ->
+                            VHLam (OptionalFoldCl opt) $ \o ->
+                            VPrim $ \s ->
+                            VPrim $ \n ->
+                            VOptionalFold a opt o s n
+    OptionalBuild    -> VPrim $ \ ~a -> VPrim $ \case
+                          VHLam (OptionalFoldCl x) _ -> x
+                          VPrimVar -> VOptionalBuild a VPrimVar
+                          t -> t `vApp` VOptional a
+                                 `vApp` VHLam (Typed "a" a) VSome
+                                 `vApp` VNone a
+
+    Record kts       -> VRecord (Dhall.Map.sort (evalE <$> kts))
+    RecordLit kts    -> VRecordLit (Dhall.Map.sort (evalE <$> kts))
+    Union kts        -> VUnion (Dhall.Map.sort ((evalE <$>) <$> kts))
+    UnionLit k v kts -> VUnionLit k (evalE v) (Dhall.Map.sort ((evalE <$>) <$> kts))
+    Combine t u      -> vCombine (evalE t) (evalE u)
+    CombineTypes t u -> vCombineTypes (evalE t) (evalE u)
+    Prefer t u       -> case (evalE t, evalE u) of
+                          (VRecordLit m, u) | null m -> u
+                          (t, VRecordLit m) | null m -> t
+                          (VRecordLit m, VRecordLit m') ->
+                             VRecordLit (Dhall.Map.sort (Dhall.Map.union m' m))
+                          (t, u) -> VPrefer t u
+    Merge x y ma     -> case (evalE x, evalE y, evalE <$> ma) of
+                          (VRecordLit m, VUnionLit k v _, _)
+                            | Just f <- Dhall.Map.lookup k m -> f `vApp` v
+                            | otherwise -> error "eval: impossible"
```

Comment:
Instead of returning `error`, would it be possible to return the fallback case (i.e. `VMerge x y ma`)?  Then the property test in `Dhall.Test.QuickCheck` could be re-enabled

Same comment for other uses of `error`

Reply:
I prefer not being silent about type errors in evaluation. This is an important point for detecting obscure bugs in type checking. Also, once we compute past an ill-typed expressions, totality and soundness of conversion checking fail to hold.

BTW, I see now that `Import.Types` has an `InternalError` exception. Should I use that instead of these ad-hoc "impossible" errors?  In this case I do intend to signal internal error, and not something like an obviously impossible branch. In Agda though, these are also called "impossible" errors.

Reply:
Yeah, you can use the `Dhall.Core.internalError` function for this purpose.  It provides a standard template for users to report implementation errors

Reply:
However, I like `error` because it gives me line numbers, while `throw InternalError` doesn't. Is there a way to get line numbers with thrown exceptions?

Reply:
Or just use `error` with a nicer message...

Reply:
I used `error` with nicer message in new commit.